lib/SILGen/SILGenPattern.cpp

```diff
@@ -2708,6 +2708,76 @@ void SILGenFunction::emitSwitchStmt(SwitchStmt *S) {
     return;
   }
 
+  // Optimize string switch statements by creating a cache that is searched through.
+  // TODO: how should I do the following? I couldn't find a StringType variable or similar.
+  if (subjectTy->getString() == "String") {
```

Comment:
Why should this be done in SILGen?

Reply:
I am not very familiar with the swift codebase. I tried a few things in a few places, and this seemed to be the best. Is there a better place for it?

Reply:
There are 2 possibilities: in SILGen and as an optimization pass.
With an optimization pass you would also catch code, which does not stem from switch statements, e.g. a series of if-statements with string compares. Also, SILGen is already a pretty big component and everything we can move out of SILGen helps to keep the code more maintainable.
On the other hand,  in an optimization pass you would need to do more pattern matching to detect the code pattern. This is problematic because it's easy to get something wrong with the pattern matching and also, it's kind of fragile because if any detail changes (e.g. in the String implementation), the pattern matching may fail.

Reply:
What do you recommend? Seems like the optimization pass is better. I would love some help getting started with the pattern matching, maybe you can point me to the right file/give an example?

Reply:
I think your approach in SILGen is good for now.
Some SILGen expert, like @gottesmm, should review the code in detail.

Reply:
@eeckstein I disagree. We are actually creating new statements in SILGen. That is something we shouldn't do. SILGen wants to treat the AST as if it is walking it in a readonly way. If we want to do this sort of thing, I think we would need to do it in the type checker or something.

If we are going to do this in SILGen the implementation needs to be completely different and will involve changes to switch emission. If we do a pass, I think it will be simpler and TBH with proper testing, I don't think it will be an issue since we should be able to make a -Onone test that pattern matches this directly.

Reply:
Do a git grep in SILGen for "::create" you will see there are no cases currently in SILGen where we are doing this.

Reply:
It would seem odd to me that it would go in the type checker (but I also don't know much about the swift codebase). @gottesmm could you point me in the direction of another place something similar is being done? Maybe it would be good to look at other optimizations and do a similar thing.

Reply:
That was my point. It would break layering to create nodes here and not in the type checker, yet would be weird to have the type checker synthesize this sort of code. That suggests that rather than creating nodes here and just running the already written code one would have to modify the underlying switch emission to understand how to do this peephole.

In terms of the type checker synthesizing code, it does it a bunch. Especially with derived conformances, creating inits, etc.

Reply:
> yet would be weird to have the type checker synthesize this sort of code. 
Does that mean I should do it elsewhere? Where should this code go? It seems like the current place is not good which leaves the type checker or SIL optimizer.

Reply:
No, I am saying you have two choices here:

1. Change the switch emission code to just recognize the pattern. This would not involve creating new nodes. Instead it would be an analysis of the current nodes.
2. Do it in the SILOptimizer.

Reply:
Alright. Thank you for clarifying :) What would you recommend? As I have said before, my knowledge of the swift codebase is limited. Which would be most straightforward and most similar to other optimizations? Thanks again for all the help.

Reply:
@gottesmm ok, makes sense

Reply:
I have started adding an optimization pass in SILCodeMotion. How should I create a call to a function (the caching function)? Is there a way to do that with `ApplyInst` or something similar?